### PR TITLE
(PDB-2634) Hash projection for (some) json fields

### PR DESCRIPTION
**Pull request by austb**
See commit message.

It's ready!

### CODE REVIEW

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 1]
```diff
@@ -1195,6 +1202,18 @@
...
+(defn projectable-json-fields
+  "Returns a list of fields of queryable json fields whose
+   field type supports json subtree extraction.
+
+   Currently only queryable json fields whose :field value
+   is a keyword or raw SQL support json subtree extraction."
+  [{:keys [projections]}]
+  (->> projections
+       (filter (comp #(and (= :queryable-json (:type %))
+                           (:projectable-json? %)) val))
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: Think val should be on a separate line so it is more obviously a comp argument than an additional and clause.

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 2]
```diff
@@ -1775,13 +1800,37 @@
...
+   :queryable? false
+   :field (hcore/raw
+            (str
+              ; Extract the original field as a string
+              (->> json-path
+                   first
+                   projections
+                   stringify-field)
+              ; append any required ->'key' operators
+              (when (> (count json-path) 1)
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: Could say `(pos? (count json-path))`, but think it's more idomatic to say `(seq json-path)`.

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 3]
```diff
@@ -1775,13 +1800,37 @@
...
 
+(defn create-json-subtree-projection
+  [dotted-field projections]
+  (let [json-path (str/split dotted-field #"\.")
+        stringify-field (fn [{:keys [field field-type]}]
+                          (case field-type
+                            :keyword (name field)
+                            :raw (:s field)
+                            (throw (IllegalArgumentException.
+                                           (tru "Cannot determine field type for field ''{0}''" field)))))]
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: Indent is a bit off.

File: src/puppetlabs/puppetdb/query_eng/engine.clj (hunk 2 above)
Comment (austb) [reply src/puppetlabs/puppetdb/query_eng/engine.clj]: I think that would only work for `> 0`, here I need to check that  there are actually json keys to look up.

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 4]
```diff
@@ -1102,6 +1108,7 @@
                  :queryable? true
                  :field :timestamp}
     "facts" {:type :queryable-json
+             :projectable-json? false
```
[the remark this answers is not in the document]
Comment (austb) [reply src/puppetlabs/puppetdb/query_eng/engine.clj]: I don't feel too strongly about this, it'll work either way. I was just marking that as "does NOT support sub-projection". Let me know if you think it should be dropped.

**Pull request merged**